les/peer.go

```diff
@@ -585,102 +593,93 @@ func (p *peer) Handshake(td *big.Int, head common.Hash, headNum uint64, genesis
 		return err
 	}
-	if err := recv.get("networkId", &rNetwork); err != nil {
-		return err
+	if int(remoteVersion) != p.version {
+		return errResp(ErrProtocolVersionMismatch, "%d (!= %d)", remoteVersion, p.version)
 	}
-	if err := recv.get("headTd", &rTd); err != nil {
+	// Check and compare the network id of remote peer
+	var remoteNetwork uint64
+	if err := recv.get("networkId", &remoteNetwork); err != nil {
 		return err
 	}
-	if err := recv.get("headHash", &rHash); err != nil {
-		return err
+	if remoteNetwork != p.network {
+		return errResp(ErrNetworkIdMismatch, "%d (!= %d)", remoteNetwork, p.network)
 	}
-	if err := recv.get("headNum", &rNum); err != nil {
+	// Check and compare the genesis of remote peer
+	var remoteGenesis common.Hash
+	if err := recv.get("genesisHash", &remoteGenesis); err != nil {
 		return err
 	}
-	if err := recv.get("genesisHash", &rGenesis); err != nil {
+	if remoteGenesis != genesis {
+		return errResp(ErrGenesisBlockMismatch, "%x (!= %x)", remoteGenesis[:8], genesis[:8])
+	}
+	// Check forkID if the protocol version is beyond the les4
+	if p.version >= lpv4 {
+		var forkID forkid.ID
+		if err := recv.get("forkID", &forkID); err != nil {
+			return err
+		}
+		if err := forkFilter(forkID); err != nil {
+			return errResp(ErrForkIDRejected, "%v", err)
+		}
+	}
+	// Pass all checks, extract the remaning fields.
+	var remoteId *big.Int
+	if err := recv.get("headTd", &remoteId); err != nil {
 		return err
 	}
-
-	if rGenesis != genesis {
-		return errResp(ErrGenesisBlockMismatch, "%x (!= %x)", rGenesis[:8], genesis[:8])
+	var remoteHead common.Hash
+	if err := recv.get("headHash", &remoteHead); err != nil {
+		return err
 	}
-	if rNetwork != p.network {
-		return errResp(ErrNetworkIdMismatch, "%d (!= %d)", rNetwork, p.network)
+	var remoteHeadNum uint64
+	if err := recv.get("headNum", &remoteHeadNum); err != nil {
+		return err
 	}
-	if int(rVersion) != p.version {
-		return errResp(ErrProtocolVersionMismatch, "%d (!= %d)", rVersion, p.version)
+	p.headInfo = &announceData{Hash: remoteHead, Number: remoteHeadNum, Td: remoteId}
+	if recvCallback != nil {
+		return recvCallback(recv)
 	}
+	return nil
+}
 
-	if server != nil {
-		if recv.get("announceType", &p.announceType) != nil {
-			// set default announceType on server side
-			p.announceType = announceTypeSimple
+// handshakeWithServer executes the les protocol handshake with les server, negotiating
+// version number, network IDs, difficulties, head and genesis blocks.
+func (p *peer) handshakeWithServer(td *big.Int, head common.Hash, headNum uint64, genesis common.Hash, forkID forkid.ID, forkFilter forkid.Filter) error {
+	return p.handshake(td, head, headNum, genesis, forkID, forkFilter, func(lists *keyValueList) {
+		// Add some client-specific handshake fields
+		//
+		// Enable signed announcement randomly even the server is not trusted.
+		p.announceType = announceTypeSimple
+		if p.trusted || r.Intn(10) > 3 {
```

Comment:
This looks a bit weird. Maybe it is justified but I don't yet understand the reasoning behind this change. What difference does it make for the server if the client has selected it as trusted? If this is a problem for security reasons (I'm not sure if it is indeed) then we can also just always request signed headers, it is not a big overhead. I'm still sceptical though because trusted servers are operated by well-known entities and their operators know anyways that their server is usually trusted by clients. Especially if it is an "announce only" server which cannot be used for anything else.

Reply:
The reason for the change is for security. Since now the server "can know" that he is selected as a trusted server since we won't request signed header for "untrusted" server right now. For the security perspective, it's may not a good thing. 

So the change will select a few random "untrusted" server for signed header too. So it's hard to judge whether they are selected.

But you are right, maybe (1) we can choose to request signed header always (2) If the trusted server is basically based on the off-chain reputation so that it's totally another thing.

But I think we definitely can find some random server as the trusted instead of following "well-known" nodes. Is it safer?